syntax = "proto3";

package test.server.config_validation;

import "validate/validate.proto";

message AddListener {
  uint32 listener_num = 1;
  uint32 route_num = 2;
}

message AddRoute {
  uint32 route_num = 1;
}

message RemoveListener {
  uint32 listener_num = 1;
}

message RemoveRoute {
  uint32 route_num = 1;
}

message Action {
  oneof action_selector {
    option (validate.required) = true;

    AddListener add_listener = 1;
    AddRoute add_route = 2;
    RemoveListener remove_listener = 3;
    RemoveRoute remove_route = 4;
  }
}

message Config {
  enum IpVersion {
    IPv4 = 0;
    IPv6 = 1;
  }
  enum ClientType {
    ENVOY_GRPC = 0;
    GOOGLE_GRPC = 1;
  }
  enum SotwOrDelta {
    SOTW = 0;
    DELTA = 1;
  }
  IpVersion ip_version = 1;
  ClientType client_type = 2;
  SotwOrDelta sotw_or_delta = 3;
}

message XdsTestCase {
  repeated Action actions = 1;
  Config config = 2;
}
